tools/template_based_query_generation/src/main/java/Node.java

```diff
@@ -0,0 +1,99 @@
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.TreeMap;
+import java.util.Random;
+
+/**
+ * class representing node for a query in the markov chain
+ */
+public class Node<E> {
+
+    private E obj;
+    private HashMap<Node<E>, Double> neighbors;
```

Comment:
We still store duplicated data in this neighbors hashmap and the following cumulativeProbabilities. Even though the storage space is not sensitive in this project, storing same info in different places and also expect those 2 objects storing identical nodes could possible cause more errors in the future when we or someone else edit it. Do we really need to store this neighbor variable? Looks like the probability was only used to calculated the cumulativeProbability. If the probability need to be kept for some other use, can we store it in the cumulativeProbabilities object? For example, use TreeMap<Double, Pair<Node<E>, Double>>

Reply:
I think neighbors would come in handy if we ever needed to modify the markov chain. However, from how we imagine people using our tool, the markov chain will be initialized and used to generate queries and won't be modified. The user will modify the config and recreate a markov chain to make modifications. So I think it's ok to completely get rid of neighbors.

Reply:
I kept neighbors because it was nice having it for testing (though it's never used by the other classes). I got rid of it and made edits to NodeTest to only use cumulativeProbabilities.